Fixed memory leak in Buffer::waitForTransform

marting87 opened the pull request:
We found that the memory usage of nav2_amcl was increasing in time. After a couple of days running this resulted in a crash. By simulating a high-speed laser-scan (150Hz) we were able to speed-up the process and trace the leak.

Amcl is using a MessageFilter on the laser_scan (message_filter.h).  On the following line, a _waitForTransform_ is called:

```
auto future = buffer_.waitForTransform(
            target_frame,
            frame_id,
            tf2::timeFromSec(stamp.seconds()),
            buffer_timeout_,
            std::bind(&MessageFilter::transformReadyCallback, this, std::placeholders::_1, next_handle_index_));
```
In waitForTransform ([tf2_ros/src/buffer.cpp](https://github.com/ros2/geometry2/blob/eloquent/tf2_ros/src/buffer.cpp)) a call is made to **addTransformableCallback()** and **addTransformableRequest()** in [tf2/src/buffer_core.cpp](https://github.com/ros2/geometry2/blob/eloquent/tf2/src/buffer_core.cpp). These functions are adding to the unordered_map **transformable_callbacks_** and vector **transformable_requests_**. 

Elements of transformable_callbacks_ are never removed. Resulting an a growing list, eating up the memory. 

![fix2](https://user-images.githubusercontent.com/22637552/86006824-d55e5f00-ba16-11ea-9458-239a21003ab2.png)

The fix below is not the prettiest, but it does the job. I'm happy to iterate on this. 

### Related issues:
- https://github.com/ros2/geometry2/issues/213
- https://github.com/ros2/geometry2/issues/200

clalancette reviewed (commented):
First, thanks for tracking down this leak and giving a good description of the problem.

As far as the implementation goes, I have some concerns with it:
1.  It seems like the vector of `callback_handles_` could get long, and hence iterating over it could get expensive.
1.  In some sense, doing the cleanup at the beginning of `waitForTransform` is giving work to the next caller that they didn't initiate and didn't ask for.  I guess for most use-cases this will likely be the same thread, but I don't love doing it this way.
1.  It seems like the tracking of the `TransformableCallbackHandle` in both the unordered_map and the vector is redundant.

Here's one idea that would solve most of the above: instead of cleaning up those TransformableCallbackHandles at the beginning of `waitForTransform`, what if we instead cleaned them up at the end of the `timerCallback` where they were used?

marting87 commented:
In waitForTransform we are also erasing from the unordered_map in the actual TransformableCallback

`this->timer_to_request_map_.erase(it->first);`

In this case we need to self-destruct the transformable callback somehow and remove it from the transformable_callbacks_. Any idea how? I cannot call removeTransformableCallback from:
```

 auto cb_handle = addTransformableCallback([this, promise, callback, future](
    tf2::TransformableRequestHandle request_handle, const std::string& target_frame,
    const std::string& source_frame, tf2::TimePoint time, tf2::TransformableResult result)
   { .. }
```

Since addTransformableCallback has alread a lock on _transformable_callbacks_mutex__;

marting87 commented:
The solution seems to be more simple then I thought. It requires an additional tf2/src/buffer_core.cpp

pschalkwijk commented:
I think I have troubles with the same buffer; when using a 15Hz laserscan memory grows with about 500 MB per day, until the system crashes. The first fix you proposed changes this to about 600MB in a week.

mdtoom commented:
Confirming that I also seem to have this issue.

clalancette commented on tf2_ros/src/buffer.cpp:
```diff
@@ -242,10 +242,12 @@ Buffer::waitForTransform(const std::string& target_frame, const std::string& sou
     // Never transformable
     promise->set_exception(std::make_exception_ptr<tf2::LookupException>(
           "Failed to transform from " + source_frame + " to " + target_frame));
+    removeTransformableCallback(cb_handle);
```
These look good to me.  However, I think there is also one more case missing, in the `timerCallback`.  As I understand it, if the timer expires before the future is completed, then the timer will go and call `cancelTransformableRequest`.  At that point, it *also* needs to call `removeTransformableCallback`, otherwise we will have a leak there, I believe.

marting87 replied:
I agree. I believe we only need to call `removeTransformavleCallback`, since it iterates over the transformable requests and removes all associated requests.

clalancette reviewed (changes requested):
All right, looking good.  One more place that I *think* we need to cleanup with `removeTransformableCallback`, and then I'll run CI on this.

pschalkwijk commented on tf2_ros/src/buffer.cpp:
```diff
@@ -213,7 +213,7 @@ Buffer::waitForTransform(const std::string& target_frame, const std::string& sou
         // Check if a timeout already occurred
         for (auto it = timer_to_request_map_.begin(); it != timer_to_request_map_.end(); ++it) {
-          if (request_handle == it->second) {
+          if (request_handle == it->second.request_handle) {
```
This should be it->second.cb_handle, this will probably fix the CI failure as well
```suggestion
          if (request_handle == it->second.cb_handle) {
```

This is the only placethe name `request_handle` is used, maybe this could be changed to `cb_handle` for clarity?

marting87 replied:
request_handle is of type **tf2::TransformableRequestHandle**, cb_handle of **tf2::TransformableCallbackHandle**. I don't think we can compare them to each other.

marting87 commented:
@clalancette I think everything should be OK now. Not sure why CI failed..

marting87 commented:
@clalancette, If you require any further assistance on this PR, let me know.

clalancette commented:
@marting87 I'm sorry, this just totally fell off of my radar.  Thanks for pinging.

One thing I noticed when looking over this again is that we need to target this at the `ros2` branch, not the `eloquent` branch.  I've now changed the target, but it looks like there are now a bunch of conflicts.  Can you rebase this onto the `ros2` branch?  Once that is done, I'll run CI on it.  Once we have it on the main branch of geometry2, we can then look at backporting it to Foxy and Eloquent.

Sorry again for the delay and the additional work here.

marting87 committed:
Clear callback from transformable_callbacks_ when callback is done or not needed

marting87 commented:
@clalancette I've updated the code, should be good now. I'll create backport requests once the fix is approved.

clalancette commented:
> @clalancette I've updated the code, should be good now. I'll create backport requests once the fix is approved.

Great, thanks.  I've also rebased it today to resolve conflicts with some other PRs.

clalancette committed:
Make BufferCore callback handle maintenance internal.

That is, there is really no reason for external users to know
about TranformableCallbackHandle at all.  Instead, hide all of
the implementation detail about the handles within BufferCore
itself.  This gets simplifies the API, and allows us to ensure
that cleanup happens properly.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
@marting87 I spent a bunch of time really going into tf2 today to try and understand this.

In the end, what I came up with is that solving it at the `tf2_ros` layer is probably the wrong solution.  That is, the TransformableCallbackHandles are really an internal implementation detail of the `tf2` layer.  Because of that, the resource management should really be done there.  With that in mind, I came up with 341eaa5, which does exactly that.  It seems to work OK in my testing, but I'd appreciate a look and test on your part to see if it really solves the leaking issue.

One potential problem with that solution is that it is changing/removing public APIs.  However, those APIs are marked "internal" anyway, so I think it is OK to change them.  Pinging @tfoote for thoughts on it.

marting87 commented:
@clalancette https://github.com/ros2/geometry2/commit/341eaa5eb8fbf5cdd2089b5133d1e2f381ad1eac works for me!

clalancette commented:
> @clalancette [341eaa5](https://github.com/ros2/geometry2/commit/341eaa5eb8fbf5cdd2089b5133d1e2f381ad1eac) works for me!

Awesome, thanks for the testing.  Here's CI on this change:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=11869)](http://ci.ros2.org/job/ci_linux/11869/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6919)](http://ci.ros2.org/job/ci_linux-aarch64/6919/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=9671)](http://ci.ros2.org/job/ci_osx/9671/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=11810)](http://ci.ros2.org/job/ci_windows/11810/)

clalancette commented:
Approved, so merging.  Thanks @marting87 for the patch.

ghost commented:
@clalancette  Any idea when the backport for foxy would be done ? 
@marting87 Could these changes directly work with the foxy branch derived from this particular commit ? Update: Seems to be working thanks !

clalancette commented:
> @clalancette Any idea when the backport for foxy would be done ?

Please feel free to open up a backport PR to foxy; we can then review it and get it merged.  Thanks.
